galpy/df/quasiisothermaldf.py

```diff
@@ -1685,6 +1688,194 @@ def sampleV(self,R,z,n=1):
+        outlier_coord_v= numpy.empty((outliers_R.size, 3))
+        for i in range(outliers_R.size):
+            outlier_coord_v[i]= self.sampleV(outliers_R[i], outliers_z[i])[0]
+        #Prepare for optimizing maxVT on a grid
+        #Edges of grid determined by input, unless prespecified by users
+        if R_min is None:
+            R_min= numpy.min(normal_R)
+        if R_max is None:
+            R_max= numpy.max(normal_R)
+        if z_max is None:
+            z_max= numpy.max(normal_z)
+        z_min= 0. #Always start grid at z=0 for stars close to plane
+        #Get the new hash of the parameters of grid
+        new_hash= hashlib.md5(numpy.array([R_min,R_max,z_max,R_pixel,z_pixel])).hexdigest()
+        #Reuse old interpolated object if new hash matches the old one
+        if new_hash == self._maxVT_hash:
+            ip_max_vT= self._maxVT_ip
+        #Generate a new interpolation object if different from before
+        else:
+            R_number= int((R_max - R_min)/R_pixel)
+            z_number= int((z_max - z_min)/z_pixel)
+            R_linspace= numpy.linspace(R_min, R_max, R_number)
+            z_linspace= numpy.linspace(z_min, z_max, z_number)
+            Rv, zv= numpy.meshgrid(R_linspace, z_linspace)
+            grid= numpy.dstack((Rv, zv)) #This grid stores (R,z) coordinate
+            #Grid is a 3 dimensional array since it stores pairs of values, but 
+            #grid max vT is a 2 dimensinal array
+            grid_max_vT= numpy.empty((grid.shape[0], grid.shape[1]))
+            #Optimize max_vT on the grid
+            for i in range(z_number):
+                for j in range(R_number):
+                    R, z= grid[i][j]
+                    grid_max_vT[i][j]= optimize.fmin_powell((lambda x: -self(
+                            R,0.,x,z,0.,log=True, use_physical=False)),1.)
+            #Determine degree of interpolation
+            ky= numpy.min([R_number-1,3])
+            kx= numpy.min([z_number-1,3])
+            #Generate interpolation object
+            ip_max_vT= interpolate.RectBivariateSpline(z_linspace,R_linspace,
+                                                       grid_max_vT,kx=kx,ky=ky)
+            #Store interpolation object
+            self._maxVT_ip= ip_max_vT
+            #Update hash of parameters
+            self._maxVT_hash= new_hash
+        #Evaluate interpolation object to get maxVT at the normal coordinates
+        normal_max_vT= ip_max_vT.ev(normal_z, normal_R)
+        #Sample all 3 velocities at a normal point and use interpolated vT
+        normal_coord_v= self.sampleV_preoptimized(normal_R,normal_z,normal_max_vT)
+        #Combine normal and outlier result, preserving original order
+        coord_v[mask]= outlier_coord_v
+        coord_v[~mask]= normal_coord_v
+        if _APY_UNITS and self._voSet:
+            return units.Quantity(coord_v,unit=units.km/units.s)
+        else:
+            return coord_v
+
+    def sampleV_preoptimized(self,R,z,maxVT):
```

Comment:
Should this function be an internal function, starting with ``_``? Then we could remove the units (see comment below).